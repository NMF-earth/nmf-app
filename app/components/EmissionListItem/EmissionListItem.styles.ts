import { StyleSheet, ViewStyle } from "react-native";

import { Layout } from "constant";
import { Colors } from "style";

const CIRCLE_WIDTH = 34;

const iconBackground: ViewStyle = {
  position: "absolute",
  borderRadius: CIRCLE_WIDTH / 2,
  width: CIRCLE_WIDTH,
  height: CIRCLE_WIDTH,
};

export default StyleSheet.create({
  container: {
    flex: 1,
    flexDirection: "row",
    alignItems: "center",
    justifyContent: "space-between",
    marginVertical: 6,
  },
  textContainer: {
    flex: 1,
    marginLeft: Layout.PADDING_HORIZONTAL * 2 + CIRCLE_WIDTH,
  },
  icon: {
    marginHorizontal: Layout.PADDING_HORIZONTAL,
  },
  iconContainer: {
    position: "absolute",
    marginLeft: Layout.PADDING_HORIZONTAL / 2,
    marginRight: Layout.PADDING_HORIZONTAL / 2,
    justifyContent: "center",
    alignItems: "center",
  },
  detailsContainer: {
    marginVertical: 2,
    flexDirection: "row",
  },
  mitigatedCircle: {
    ...iconBackground,
    backgroundColor: Colors.primary10,
  },
  notMitigatedCircle: {
    ...iconBackground,
    backgroundColor: Colors.grey + "60",
  },
});
