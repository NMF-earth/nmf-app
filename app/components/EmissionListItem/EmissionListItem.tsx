import React from "react";
import { View, TouchableOpacity } from "react-native";
import { Ionicons } from "@expo/vector-icons";
import { FormattedNumber } from "react-native-globalize";

import { Colors } from "style";
import { RecurringEmission, Emission, EmissionType } from "interfaces";

import Text from "../Text";
import styles from "./EmissionListItem.styles";

interface AnyEmission extends RecurringEmission, Emission {}

interface EmissionListItemProps extends Partial<AnyEmission> {
  title: string;
  co2value: number;
  iconName: string;
  onPress: () => void;
  /* no used in EmissionListItem so optional */
  emissionType?: EmissionType;
  value?: number;
  creationDate?: string;
  times?: number;
}

const EmissionListItem: React.FC<EmissionListItemProps> = ({
  isMitigated,
  name = "",
  iconName = "md-car",
  title = "",
  co2value = 0,
  onPress,
}) => (
  <TouchableOpacity onPress={onPress} style={styles.container}>
    <View style={styles.iconContainer}>
      <View style={isMitigated ? styles.mitigatedCircle : styles.notMitigatedCircle} />
      <Ionicons
        name={iconName as keyof typeof Ionicons.glyphMap}
        size={22}
        style={styles.icon}
        color={isMitigated ? Colors.green50 : Colors.grey70}
      />
    </View>
    <View style={styles.textContainer}>
      <Text.Primary numberOfLines={1}>{name.length ? name : title}</Text.Primary>
      <View style={styles.detailsContainer}>
        <Text.Tertiary numberOfLines={1} lightGray>
          <FormattedNumber maximumFractionDigits={co2value >= 1 ? 2 : 4} value={co2value} /> kgCO2
        </Text.Tertiary>
        {isMitigated && (
          <>
            <Text.Tertiary lightGray>{" â€¢ "}</Text.Tertiary>
            <Text.Tertiary green>{"Offset"}</Text.Tertiary>
          </>
        )}
      </View>
    </View>
    <Ionicons
      name={"ios-chevron-forward-outline"}
      size={18}
      style={styles.icon}
      color={Colors.grey70}
    />
  </TouchableOpacity>
);

export { EmissionListItem, EmissionListItemProps };
