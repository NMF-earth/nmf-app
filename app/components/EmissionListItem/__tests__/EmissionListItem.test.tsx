import React from "react";
import renderer from "react-test-renderer";
import { EmissionListItem } from "../EmissionListItem";
import { FormattedProvider } from "react-native-globalize";

jest.unmock("../EmissionListItem.tsx");

const props = {
  id: "123",
  title: "170 g of red meat",
  co2value: 2.1,
  onPress: () => {
    // do nothing.
  }
};

it("EmissionListItem renders correctly by default", () => {
  const tree = renderer.create(
    <FormattedProvider locale="en">
      <EmissionListItem {...props} />
    </FormattedProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it("EmissionListItem renders correctly with food icon", () => {
  const tree = renderer.create(
    <FormattedProvider locale="en">
      <EmissionListItem {...props} food />
    </FormattedProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it("EmissionListItem renders correctly with transport icon", () => {
  const tree = renderer
  .create(
    <FormattedProvider locale="en">
      <EmissionListItem {...props} transport />
    </FormattedProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it("EmissionListItem renders correctly with custom icon", () => {
  const tree = renderer.create(
    <FormattedProvider locale="en">
      <EmissionListItem {...props} custom />
    </FormattedProvider>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});
