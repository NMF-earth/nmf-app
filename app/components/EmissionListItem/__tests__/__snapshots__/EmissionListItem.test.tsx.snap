// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EmissionListItem renders correctly by default 1`] = `
<FormattedProvider
  locale="en"
>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 4,
      }
    }
  >
    <Ionicons
      color="#111714"
      name=""
      size={28}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text.Primary
        numberOfLines={1}
      >
        170 g of red meat
      </Text.Primary>
      <Text.Tertiary
        light={true}
        lightGray={true}
        numberOfLines={1}
      >
        <FormattedNumber
          maximumFractionDigits={2}
          value={2.1}
        />
         
        kg
      </Text.Tertiary>
    </View>
    <Ionicons
      color="#111714"
      name="ios-arrow-forward"
      size={22}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
  </TouchableOpacity>
</FormattedProvider>
`;

exports[`EmissionListItem renders correctly with custom icon 1`] = `
<FormattedProvider
  locale="en"
>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 4,
      }
    }
  >
    <Ionicons
      color="#111714"
      name="md-build"
      size={26}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text.Primary
        numberOfLines={1}
      >
        170 g of red meat
      </Text.Primary>
      <Text.Tertiary
        light={true}
        lightGray={true}
        numberOfLines={1}
      >
        <FormattedNumber
          maximumFractionDigits={2}
          value={2.1}
        />
         
        kg
      </Text.Tertiary>
    </View>
    <Ionicons
      color="#111714"
      name="ios-arrow-forward"
      size={22}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
  </TouchableOpacity>
</FormattedProvider>
`;

exports[`EmissionListItem renders correctly with food icon 1`] = `
<FormattedProvider
  locale="en"
>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 4,
      }
    }
  >
    <Ionicons
      color="#111714"
      name="md-restaurant"
      size={28}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text.Primary
        numberOfLines={1}
      >
        170 g of red meat
      </Text.Primary>
      <Text.Tertiary
        light={true}
        lightGray={true}
        numberOfLines={1}
      >
        <FormattedNumber
          maximumFractionDigits={2}
          value={2.1}
        />
         
        kg
      </Text.Tertiary>
    </View>
    <Ionicons
      color="#111714"
      name="ios-arrow-forward"
      size={22}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
  </TouchableOpacity>
</FormattedProvider>
`;

exports[`EmissionListItem renders correctly with transport icon 1`] = `
<FormattedProvider
  locale="en"
>
  <TouchableOpacity
    onPress={[Function]}
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 4,
      }
    }
  >
    <Ionicons
      color="#111714"
      name="md-airplane"
      size={30}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <Text.Primary
        numberOfLines={1}
      >
        170 g of red meat
      </Text.Primary>
      <Text.Tertiary
        light={true}
        lightGray={true}
        numberOfLines={1}
      >
        <FormattedNumber
          maximumFractionDigits={2}
          value={2.1}
        />
         
        kg
      </Text.Tertiary>
    </View>
    <Ionicons
      color="#111714"
      name="ios-arrow-forward"
      size={22}
      style={
        Object {
          "marginTop": 2,
          "paddingHorizontal": 14,
        }
      }
    />
  </TouchableOpacity>
</FormattedProvider>
`;
